lib/spack/spack/spec.py

```diff
@@ -1642,6 +1642,7 @@ def to_dict(self, hash=ht.dag_hash):
         for s in self.traverse(order='pre', deptype=hash.deptype):
             node = s.to_node_dict(hash)
             node[s.name]['hash'] = s.dag_hash()
+            node[s.name]['full_hash'] = s._full_hash
```

Comment:
Sorry @scheibelp this is where I should have put this question originally.  

Should this be conditional, similar to build_hash, just below? Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

Reply:
> Should this be conditional, similar to build_hash, just below?

It should not be: you always want the full hash available in the dictionary. Also this won't affect the other hashes. The build hash does not need to be computed conditionally either (although that's more of an aside).

One thing I don't understand yet: I don't see the `_full_hash` actually being calculated anywhere (which is why I'm not surprised when you say it is `None`) - I would have expected that you'd call `s.full_hash()` here (which computes the hash if it hasn't been already: `_full_hash` is a "caching" variable to avoid recomputing it).

Reply:
I see your point about why calling `full_hash()` might needed here, rather than using the cached property.  The thing I was most worried about was taking a concrete spec that was reconstituted from a `spec.yaml` on a mirror, and recomputing that hash locally, which would compute it based on the local `package.py`.  But that would be when reading from the dictionary index or `spec.yaml` (calling `from_dict`), rather than writing to them (and calling `to_dict`).

However, in my testing, it seems I'm always getting the right full hash written into the database index as well as into the buildcache index and mirror `spec.yaml` files, so I'm trying to figure out how that is working.  I'm wondering if it's just happening as a normal part of concretizing a spec.  I'm not sure yet, but I'll report back when I know what's going on here.

Reply:
So it's not during concretization that `full_hash()` is getting called, but during the install process as a part of trying to install from the binary cache.  It's also getting called a couple times from `database.py` in the `_failed_spec_path` method.  I don't understand what's going on behind the scenes there, maybe @tldahlgren can shed some light.  I tried to attach the log from a session installing `zlib` with a breakpoint set in the `full_hash()` method, but github continually failed to upload the file, so I'm pasting it below:

```
$ spack install zlib
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1460)install()
-> self._init_queue(install_deps, install_package)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1040)_init_queue()
-> spack.store.db.clear_failure(self.pkg.spec, force=True)
  /data/scott/projects/spack/lib/spack/spack/database.py(463)clear_failure()
-> if self.prefix_failure_marked(spec):
  /data/scott/projects/spack/lib/spack/spack/database.py(546)prefix_failure_marked()
-> return os.path.exists(self._failed_spec_path(spec))
  /data/scott/projects/spack/lib/spack/spack/database.py(416)_failed_spec_path()
-> '{0}-{1}'.format(spec.name, spec.full_hash()))
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1501)install()
-> if pkg_id in self.failed or spack.store.db.prefix_failed(spec):
  /data/scott/projects/spack/lib/spack/spack/database.py(532)prefix_failed()
-> return self.prefix_failure_marked(spec)
  /data/scott/projects/spack/lib/spack/spack/database.py(546)prefix_failure_marked()
-> return os.path.exists(self._failed_spec_path(spec))
  /data/scott/projects/spack/lib/spack/spack/database.py(416)_failed_spec_path()
-> '{0}-{1}'.format(spec.name, spec.full_hash()))
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
==> Installing zlib
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1591)install()
-> self._install_task(task, **kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1081)_install_task()
-> full_hash_match):
  /data/scott/projects/spack/lib/spack/spack/installer.py(234)_install_from_cache()
-> pkg, explicit, unsigned=unsigned, full_hash_match=full_hash_match)
  /data/scott/projects/spack/lib/spack/spack/installer.py(355)_try_install_from_binary_cache()
-> pkg.spec, force=False, full_hash_match=full_hash_match)
  /data/scott/projects/spack/lib/spack/spack/binary_distribution.py(1227)get_mirrors_for_spec()
-> spec_full_hash = spec.full_hash()
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
==> No binary for zlib found: installing from source
==> Using cached archive: /data/scott/projects/spack/var/spack/cache/_source-cache/archive/c3/c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1.tar.gz
==> zlib: Executing phase: 'install'
[+] /data/scott/projects/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-7.4.0/zlib-1.2.11-gcrfowf2m3tywl4fh333mer3w2q6ytha
```

Reply:
So given that information, do you still think `full_hash()` needs to be called from the `spec.to_dict()` method?  I don't think it would hurt anything from my point of view (I don't think it would recompute/clobber it on a spec read in from yaml, which would be bad), and maybe it makes sense to do it just for sake of clarity when reading the code.

Reply:
> do you still think full_hash() needs to be called from the spec.to_dict()

Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching. When you update that, I'm curious if you still see the problem you mentioned in the first comment in this particular thread:

> Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

Regarding

> I don't think it would recompute/clobber it on a spec read in from yaml, which would be bad

I agree this would be a problem and we should not actually compute the full hash but instead pull it from the yaml.

Reply:
> Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching.

Sounds good, I'll call the method here instead of accessing the cached attribute.

> I'm curious if you still see the problem you mentioned in the first comment in this particular thread:
> 
> > Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

Sorry I wasn't clear about that.  It would happen when I would do something like the following in the spack python interpreter :

``` python
from spack.spec import Spec
s = Spec('zlib')
s.concretize()
with open('/some/path/spec.yaml', 'w') as fd:
    fd.write(s.to_dict())
```

Reply:
> > Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching.
> 
> Sounds good, I'll call the method here instead of accessing the cached attribute.

Just a heads up:  I made this change and it caused over a hundred tests to start failing, so I'm now looking into what's going on here.  I've verified that the commit before passes all those tests, and it's just the change from `... = s._full_hash` to `... = s.full_hash()` that causes the failures.

Reply:
There are a lot of failures of the form:

```
AttributeError: 'X' object has no attribute 'content_hash'
```

In stack traces like the following:

```
lib/spack/spack/test/database.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lib/spack/spack/database.py:120: in converter
    return function(self, spec_like, *args, **kwargs)
lib/spack/spack/database.py:1140: in add
    self._add(spec, directory_layout, explicit=explicit)
lib/spack/llnl/util/lock.py:584: in __exit__
    if self._exit(release_fn):
lib/spack/llnl/util/lock.py:605: in _exit
    return self._lock.release_write(release_fn)
lib/spack/llnl/util/lock.py:461: in release_write
    result = release_fn()
lib/spack/llnl/util/lock.py:578: in release_fn
    return self._release_fn(type, value, traceback)
lib/spack/spack/database.py:983: in _write
    self._write_to_file(f)
lib/spack/spack/database.py:617: in _write_to_file
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:617: in <genexpr>
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:212: in to_dict
    spec_dict_list = self.spec.to_dict()['spec']
lib/spack/spack/spec.py:1645: in to_dict
    node[s.name]['full_hash'] = s.full_hash()
lib/spack/spack/spec.py:1464: in full_hash
    return self._cached_hash(ht.full_hash, length)
lib/spack/spack/spec.py:1431: in _cached_hash
    hash_string = self._spec_hash(hash)
lib/spack/spack/spec.py:1404: in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
```

As I'm digging into it, it seems like maybe `MockPackageBase` is missing some properties required to handle calling `full_hash()` on a spec...

Reply:
Or maybe it would belong in the `MockPackageMultiRepo.add_package` method?

Reply:
Adding the following to `mock_package.MockPackageBase` fixes `spack test -k database` (I didn't run all tests):

```
    def content_hash(self):
        # Unlike real packages, MockPackage doesn't have a corresponding
        # package.py file; in that sense, the content_hash is always the same.
        return self.__class__.__name__
```

Reply:
Thanks @scheibelp I did something similar which fixed maybe 70% of the failures.  But I'll replace what I did with what you have here.

Reply:
I fixed several more of the failures by guarding the call to `full_hash()` inside `to_dict()` with a check for `spec.concrete()`.  If not concrete, I used the `_full_hash` property, but maybe `None` would be a better choice?

Reply:
This sounds good but IMO not setting `'full_hash'` when `not spec.concrete` would be clearest.

Reply:
Thanks for your help, I'm down to 22 failures.  Only a few of them fail when run in isolation, passing until they're run with (specific, but as yet unknown to me) other tests.  For example 5 of the `cmd/uninstall.py` tests fail when I run all the tests, but they all pass if I just run the `cmd/uninstall.py` tests.  Ditto for 7 of the `cmd/find.py` tests.

Reply:
Thanks for pursuing this - I think this is worthwhile:

* using `.full_hash()` will reduce fragility of `to_dict`
* the tests that are failing are likely fragile